pkg/ipam/crd.go

```diff
@@ -0,0 +1,599 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipam
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/pkg/aws/metadata"
+	"github.com/cilium/cilium/pkg/cidr"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/k8s"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	k8sversion "github.com/cilium/cilium/pkg/k8s/version"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/trigger"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	sharedNodeStore *nodeStore
+	initNodeStore   sync.Once
+)
+
+const (
+	// customResourceUpdateRate is the maximum rate in which a custom
+	// resource is updated
+	customResourceUpdateRate = 15 * time.Second
+)
+
+// nodeStore represents a CiliumNode custom resource and binds the CR to a list
+// of allocators
+type nodeStore struct {
+	// mutex protects access to all members of this struct
+	mutex lock.RWMutex
+
+	// ownNode is the last known version of the own node resource
+	ownNode *ciliumv2.CiliumNode
+
+	// allocators is a list of allocators tied to this custom resource
+	allocators []*crdAllocator
+
+	// refreshTrigger is the configured trigger to synchronize updates to
+	// the custom resource with rate limiting
+	refreshTrigger *trigger.Trigger
+
+	// availableIPs is the cached version of all available IPs
```

Comment:
```suggestion
	// availableIPs maps an address family to the size of the IP allocation pool for that family.
```

`availableIPs` seems to suggest it's to do with the number in the pool that can be allocated, but this appears to actually be the IP allocation pool size.

Reply:
Renamed to `allocationPoolSize` to make it clear.